src/lib/focus-handler.js

```diff
@@ -0,0 +1,206 @@
+import * as React from "react"
+import { useFocusContext } from "./hooks-context"
+import { FocusContext } from "./hooks-create-context"
+
+export const FocusHandler = ({ uri, location, component, ...domProps }) => {
+  const requestFocus = useFocusContext()
+
+  return (
+    <FocusHandlerImpl
+      {...domProps}
+      component={component}
+      requestFocus={requestFocus}
+      uri={uri}
+      location={location}
+    />
+  )
+}
+
+const FocusHandlerImpl = ({
+  children,
+  style,
+  requestFocus,
+  component: Comp = "div",
+  uri,
+  location,
+  ...domProps
+}) => {
+  let node = React.useRef()
+
+  const [shouldFocus, setShouldFocus] = React.useState(true)
+  const [prevUri, setPrevUri] = React.useState(uri)
+  const [prevPathname, setPrevPathname] = React.useState(location.pathname)
+  const [focusHandlerCount, setFocusHandlerCount] = React.useState(0)
+  const initialRender = React.useRef(true)
+
+  if (uri) {
+    const myURIChanged = uri !== prevUri
+    const navigatedUpToMe =
+      prevPathname !== location.pathname && location.pathname === uri
+
+    setPrevUri(uri)
+    setPrevPathname(location.pathname)
+
+    setShouldFocus(myURIChanged || navigatedUpToMe)
+  }
+
+  React.useEffect(() => {
+    setFocusHandlerCount(f => f + 1)
+    focus()
+
+    return () => {
+      setFocusHandlerCount(f => f - 1)
+      if (focusHandlerCount === 0) {
+        initialRender.current = true
+      }
+    }
+  }, [])
+
+  React.useEffect(() => {
+    if (shouldFocus) {
+      focus()
+    }
+  }, [location])
+
+  const focus = () => {
+    if (process.env.NODE_ENV === "test") {
+      // TODO: Still a problem?
+      // getting cannot read property focus of null in the tests
+      // and that bit of global `initialRender` state causes problems
+      return
+    }
+
+    if (requestFocus) {
+      requestFocus(node.current)
+    } else {
+      if (initialRender.current) {
+        initialRender.current = false
+      } else if (node) {
+        // React polyfills [autofocus] and it fires earlier than cDM, so we were stealing focus away, this line prevents that.
+        if (!node.current.contains(document.activeElement)) {
+          node.current.focus()
+        }
+      }
+    }
+  }
+
+  const _requestFocus = node => {
+    if (!shouldFocus && node) {
+      node.current.focus()
+    }
+  }
+
+  return (
+    <Comp
+      style={{ outline: "none", ...style }}
+      tabIndex="-1"
+      ref={n => (node = n)}
```

Comment:
Correct use of ref
```suggestion
      ref={node}
```

Reply:
I think this was done on purpose

Reply:
https://blog.thoughtspile.tech/2021/05/17/everything-about-react-refs/

> There’s a subtle difference between having an arrow as your ref prop and a ref object or a stable callback — the arrow has a new identity on every render, forcing the ref to go through an update cycle null. This is normally not too bad, but good to know.

See ref arrows

```jsx
// this does node -> null -> node on every render
<div ref={e => this.node = e} />
// this doesn't
<div ref={useCallback(e => this.node = e, [])} />
// neither does this
setRef = e => this.node = e;
<div ref={this.setRef} />
// this is fine, too
const ref = useRef();
<div ref={ref} />
```

This was added in https://github.com/reach/router/commit/738b3caa41fca4d327aeac7123beeca56ec18554 but no explanation.

Like I knew how to correctly do this, I just wanted to keep old behavior. If we thoroughly test this with all browser, screen readers, keyboard etc. and it works with the above, too, then we can change

Reply:
I don't really understand why you want do have that behavior 😅 Cause refs are not tracked by React, there is no real benefit of putting them to null but let's keep it how it is